Robustness of vSPC to malformed data

I was attempting to determine how best to confirm connectivity between an ESX host and the endpoint and tried running the following on the host:
```
ssh <endpoint ip> -p 2377
nc -v <endpoint ip> 2377
```

The result was a vSPC crash.
```
Mar 31 2017 14:38:43.299Z INFO  connection received
Mar 31 2017 14:38:43.299Z INFO  sending WILL 3
Mar 31 2017 14:38:43.299Z DEBUG entered write loop
Mar 31 2017 14:38:43.300Z DEBUG Sending command: 251 3
Mar 31 2017 14:38:43.300Z INFO  sending WILL 1
Mar 31 2017 14:38:43.300Z DEBUG Sending command: 251 1
Mar 31 2017 14:38:43.300Z INFO  sending WILL 0
Mar 31 2017 14:38:43.300Z DEBUG Sending command: 251 0
Mar 31 2017 14:38:43.300Z INFO  sending DO 3
Mar 31 2017 14:38:43.300Z DEBUG Sending command: 253 3
Mar 31 2017 14:38:43.300Z INFO  sending DO 232
Mar 31 2017 14:38:43.300Z DEBUG Sending command: 253 232
Mar 31 2017 14:38:43.300Z INFO  sending DO 0
Mar 31 2017 14:38:43.300Z DEBUG Sending command: 253 0
Mar 31 2017 14:38:43.329Z ERROR vspc datahandler cannot find the vm with the provided data
Mar 31 2017 14:38:43.330Z DEBUG data handler closed
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xe6c842]

goroutine 1418 [running]:
github.com/vmware/vic/lib/vspc.(*handler).dataHdlr(0xc4204441f8, 0x7f87573ff1a0, 0xc420444030, 0xc42094e600, 0x2, 0x600, 0xc420fca000)
        /home/vagrant/vicsmb/src/github.com/vmware/vic/lib/vspc/handlers.go:42 +0x52
github.com/vmware/vic/lib/vspc.(*handler).(github.com/vmware/vic/lib/vspc.dataHdlr)-fm(0x7f87573ff1a0, 0xc420444030, 0xc42094e600, 0x2, 0x600, 0xc420fca000)
        /home/vagrant/vicsmb/src/github.com/vmware/vic/lib/vspc/vspc.go:135 +0x66
github.com/vmware/vic/pkg/telnet.(*Conn).dataHandlerWrapper(0xc420fca000, 0x7f87573ff1a0, 0xc420444030, 0x7f87573ff1c8, 0xc420166150)
        /home/vagrant/vicsmb/src/github.com/vmware/vic/pkg/telnet/connection.go:214 +0x183
created by github.com/vmware/vic/pkg/telnet.(*Server).Accept
        /home/vagrant/vicsmb/src/github.com/vmware/vic/pkg/telnet/server.go:95 +0x262
```

This gets a security label because it's a potential DoS vector, malicious or otherwise and could be triggered by basic port scanning.
The priority is set to high as some of our PSR requirements will involve port scanning which we won't be able to exit with a functional system.

--- a/lib/vspc/handlers.go
+++ b/lib/vspc/handlers.go
@@ -37,9 +37,16 @@ type handler struct {
 func (h *handler) dataHdlr(w io.Writer, b []byte, tc *telnet.Conn) {
 	cvm, exists := h.vspc.cvmFromTelnetConn(tc)
 	if !exists {
-		log.Errorf("vspc datahandler cannot find the vm with the provided data")
+		// the fsm will sense the closed connecrion and perform the necessary cleanup
+		if tc.UnderlyingConnection() != nil {
+			log.Errorln("cannot find a vm associated with this connection.")
+			log.Infoln("closing connection")
+			tc.UnderlyingConnection().Close()
+		}
+		return
 	}
 	cvm.remoteConn.Write(b)
+
 }
 
 // CmdHdlr is the telnet command handler

--- a/pkg/telnet/connection.go
+++ b/pkg/telnet/connection.go
@@ -125,6 +125,11 @@ func newConn(opts *connOpts) *Conn {
 	return tc
 }
 
+//UnderlyingConnection returns the underlying TCP connection
+func (c *Conn) UnderlyingConnection() net.Conn {
+	return c.conn
+}
+
 func (c *Conn) writeLoop() {
 	log.Debugf("entered write loop")
 	for {
@@ -149,10 +154,11 @@ func (c *Conn) startNegotiation() {
 	}
 }
 
-// Close closes the telnet connection
-func (c *Conn) Close() {
+// close closes the telnet connection
+func (c *Conn) close() {
 	c.closedMutex.Lock()
 	defer c.closedMutex.Unlock()
+
 	c.closed = true
 	log.Infof("Closing the connection")
 	c.conn.Close()

--- a/pkg/telnet/fsm.go
+++ b/pkg/telnet/fsm.go
@@ -56,7 +56,7 @@ func (fsm *fsm) start() {
 		}
 		if err != nil {
 			log.Debugf("connection read: %v", err)
-			fsm.tc.Close()
+			fsm.tc.close()
 			break
 		}
 	}

